From a1356951ee80cf01e444c83403fdd65f7ef8b38c Mon Sep 17 00:00:00 2001
From: Baisong Zhong <zhongbaisong@huawei.com>
Date: Fri, 31 Mar 2023 14:47:44 +0800
Subject: [PATCH 901/921] [Huawei] tc_nat: Add bpf_ct_nf_conntrack_in and
 bpf_ct_nf_conntrack_confirm kfuncs

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 188604

--------------------------------

This patch add bpf_ct_nf_conntrack_in and bpf_ct_nf_conntrack_confirm kfuncs

Signed-off-by: Baisong Zhong <zhongbaisong@huawei.com>
---
 net/netfilter/nf_conntrack_bpf.c | 29 +++++++++++++++++++++++++++++
 1 file changed, 29 insertions(+)

diff --git a/net/netfilter/nf_conntrack_bpf.c b/net/netfilter/nf_conntrack_bpf.c
index 02a9f45d202b..c9bdabc88e44 100644
--- a/net/netfilter/nf_conntrack_bpf.c
+++ b/net/netfilter/nf_conntrack_bpf.c
@@ -359,6 +359,33 @@ void bpf_ct_release(struct nf_conn *nfct)
 	nf_ct_put(nfct);
 }
 
+static inline void bpf_nf_hook_state_init(struct nf_hook_state *p, struct sk_buff *skb)
+{
+	p->hook = NF_INET_PRE_ROUTING;
+	p->pf = NFPROTO_IPV4;
+	p->in = skb->dev;
+	p->out = NULL;
+	p->sk = NULL;
+	p->net = dev_net(skb->dev);
+	p->okfn = NULL;
+}
+
+int bpf_ct_nf_conntrack_in(struct __sk_buff *skb_ctx)
+{
+	struct nf_hook_state state;
+	struct sk_buff *skb = (struct sk_buff *)skb_ctx;
+
+	bpf_nf_hook_state_init(&state, skb);
+	return nf_conntrack_in(skb, &state);
+}
+
+int bpf_ct_nf_conntrack_confirm(struct __sk_buff *skb_ctx)
+{
+	struct sk_buff *skb = (struct sk_buff *)skb_ctx;
+
+	return nf_conntrack_confirm(skb);
+}
+
 __diag_pop()
 
 BTF_SET_START(nf_ct_kfunc_ids)
@@ -368,6 +395,8 @@ BTF_ID(func, bpf_skb_ct_alloc)
 BTF_ID(func, bpf_skb_ct_lookup)
 BTF_ID(func, bpf_ct_insert_entry)
 BTF_ID(func, bpf_ct_release)
+BTF_ID(func, bpf_ct_nf_conntrack_in)
+BTF_ID(func, bpf_ct_nf_conntrack_confirm)
 BTF_SET_END(nf_ct_kfunc_ids)
 
 static const struct btf_kfunc_id_set nf_conntrack_kfunc_set = {
-- 
2.17.1

